File: PlaywrightDotNet/PlaywrightDotNet/PageObjectModel/Utilities.cs
using System.Diagnostics;

namespace PlaywrightDotNet.PageObjectModel
{
    internal static class Utilities
    {
        /// <summary>
        /// Uses a retry pattern to wait for the locator to be visible.
        /// Only use if there is true uncertainty about an element appearing.
        /// If an element is there in some environments but not others
        /// then use environment variables to determine whether or not to look for it
        /// </summary>
        /// <param name="locator"></param>
        /// <returns></returns>
        public static async Task<bool> WaitForVisibleWithoutThrowing(this ILocator locator)
        {
            const int timeout = 2000; // We need to use a relatively short timeout to ensure we don't hold up tests
            const int step = 500;

            var sw = Stopwatch.StartNew();
            await locator.Page.WaitForLoadStateAsync(LoadState.NetworkIdle);

            while (sw.ElapsedMilliseconds < timeout)
            {
                if (await locator.IsVisibleAsync())
                {
                    return true;
                }
                
                await Task.Delay(step);
            }

            return false;
        }
    }
}

File: PlaywrightTypeScript/tests/page-object-model/utilities.ts
import { Locator } from "@playwright/test";

export class Utilities {
    /**
     * Uses a retry pattern to wait for the locator to be visible.
     * 
     * Only use if there is uncertainty about an element appearing, 
     * if an element is there in some environments but not others 
     * then use environment variables to determine whether or not to look for it
     * @param {Locator} locator
     */
    static async waitForVisibleWithoutThrowing(locator: Locator): Promise<boolean> {
        let isVisible = await this.waitUntilTrueOrTimeout(() => locator.isVisible())        
        return isVisible;
    }

    static async waitUntilTrueOrTimeout(action: () => Promise<boolean>, timeout: number = 5000): Promise<boolean> {
      const maxtime = Date.now() + timeout;
      const step = 500;
    
      while (Date.now() < maxtime) {
        if (await action()) {
          return true;
        }
        else {
          await this.delay(step);
        }
      }

      return false;
    }

    private static delay(ms: number)
    {
      return new Promise(resolve => setTimeout(resolve, ms));
    }
}